From 7d2f21d50e06755968953093a5d5badc811cbbc1 Mon Sep 17 00:00:00 2001
From: Venkat Chimata <venkat@nearhop.com>
Date: Thu, 5 Feb 2026 22:13:56 +0530
Subject: [PATCH] hostapd/ubus: Clear bss_mgmt_enable params when they are
 disabled.

bss_mgmt_enable params -neighbor_report, beacon_report and bss_transition -
once enabled through the ubus, are not disabled.
Disable them when they are not set.

Signed-off-by: Venkat Chimata <venkat@nearhop.com>
---
 src/ap/ubus.c | 88 +++++++++++++++++++++++++++++++++++++++++++++------
 1 file changed, 79 insertions(+), 9 deletions(-)

diff --git a/src/ap/ubus.c b/src/ap/ubus.c
index 4eb5a11..30e8a21 100644
--- a/src/ap/ubus.c
+++ b/src/ap/ubus.c
@@ -947,7 +947,6 @@ __hostapd_bss_mgmt_enable_f(struct hostapd_data *hapd, int flag)
 		if (bss->radio_measurements[0] &
 		    WLAN_RRM_CAPS_NEIGHBOR_REPORT)
 			return false;
-
 		bss->radio_measurements[0] |=
 			WLAN_RRM_CAPS_NEIGHBOR_REPORT;
 		hostapd_neighbor_set_own_report(hapd);
@@ -973,6 +972,70 @@ __hostapd_bss_mgmt_enable_f(struct hostapd_data *hapd, int flag)
 	}
 }
 
+static bool
+__hostapd_bss_mgmt_disable_f(struct hostapd_data *hapd, int flag)
+{
+	struct hostapd_bss_config *bss = hapd->conf;
+	uint32_t flags;
+
+	switch (flag) {
+	case BSS_MGMT_EN_NEIGHBOR:
+		if (bss->radio_measurements[0] &
+		    WLAN_RRM_CAPS_NEIGHBOR_REPORT) {
+			bss->radio_measurements[0] &=
+				~WLAN_RRM_CAPS_NEIGHBOR_REPORT;
+			return true;
+		}
+		return false;
+	case BSS_MGMT_EN_BEACON:
+		flags = WLAN_RRM_CAPS_BEACON_REPORT_PASSIVE |
+			WLAN_RRM_CAPS_BEACON_REPORT_ACTIVE |
+			WLAN_RRM_CAPS_BEACON_REPORT_TABLE;
+
+		if (bss->radio_measurements[0] & flags == flags) {
+			bss->radio_measurements[0] &= ~(u8) flags;
+			return true;
+		}
+		return false;
+#ifdef CONFIG_WNM_AP
+	case BSS_MGMT_EN_BSS_TRANSITION:
+		if (bss->bss_transition) {
+			bss->bss_transition = 0;
+			return true;
+		}
+
+		bss->bss_transition = 0;
+		return false;
+#endif
+	}
+}
+
+static int
+__hostapd_bss_mgmt_get_f(struct hostapd_data *hapd)
+{
+	struct hostapd_bss_config *bss = hapd->conf;
+	uint32_t beacon_flags;
+	int flags = 0;
+
+
+	if (bss->radio_measurements[0] &
+	    WLAN_RRM_CAPS_NEIGHBOR_REPORT) {
+		flags |= (1 << BSS_MGMT_EN_NEIGHBOR);
+	}
+	beacon_flags = WLAN_RRM_CAPS_BEACON_REPORT_PASSIVE |
+			WLAN_RRM_CAPS_BEACON_REPORT_ACTIVE |
+			WLAN_RRM_CAPS_BEACON_REPORT_TABLE;
+
+	if (bss->radio_measurements[0] & beacon_flags == beacon_flags) {
+		flags |= (1 << BSS_MGMT_EN_BEACON);
+	}
+#ifdef CONFIG_WNM_AP
+	if (bss->bss_transition) {
+		flags |= (1 << BSS_MGMT_EN_BSS_TRANSITION);
+	}
+#endif
+	return flags;
+}
 static void
 __hostapd_bss_mgmt_enable(struct hostapd_data *hapd, uint32_t flags)
 {
@@ -980,10 +1043,11 @@ __hostapd_bss_mgmt_enable(struct hostapd_data *hapd, uint32_t flags)
 	int i;
 
 	for (i = 0; i < __BSS_MGMT_EN_MAX; i++) {
-		if (!(flags & (1 << i)))
-			continue;
-
-		update |= __hostapd_bss_mgmt_enable_f(hapd, i);
+		if (!(flags & (1 << i))) {
+			update |= __hostapd_bss_mgmt_disable_f(hapd, i);
+		} else {
+			update |= __hostapd_bss_mgmt_enable_f(hapd, i);
+		}
 	}
 
 	if (update)
@@ -1014,11 +1078,16 @@ hostapd_bss_mgmt_enable(struct ubus_context *ctx, struct ubus_object *obj,
 
 	blobmsg_parse(bss_mgmt_enable_policy, __BSS_MGMT_EN_MAX, tb, blob_data(msg), blob_len(msg));
 
+	flags = __hostapd_bss_mgmt_get_f(hapd);
 	for (i = 0; i < ARRAY_SIZE(tb); i++) {
-		if (!tb[i] || !blobmsg_get_bool(tb[i]))
+		if (!tb[i]) {
 			continue;
-
-		flags |= (1 << i);
+		}
+		if (blobmsg_get_bool(tb[i])) {
+			flags |= (1 << i);
+		} else {
+			flags &= ~(1 << i);
+		}
 	}
 
 	__hostapd_bss_mgmt_enable(hapd, flags);
@@ -1028,7 +1097,8 @@ hostapd_bss_mgmt_enable(struct ubus_context *ctx, struct ubus_object *obj,
 static void
 hostapd_rrm_nr_enable(struct hostapd_data *hapd)
 {
-	__hostapd_bss_mgmt_enable(hapd, 1 << BSS_MGMT_EN_NEIGHBOR);
+	int flags = __hostapd_bss_mgmt_get_f(hapd);
+	__hostapd_bss_mgmt_enable(hapd, flags | (1 << BSS_MGMT_EN_NEIGHBOR));
 }
 
 static int
-- 
2.34.1

